src/ITensorNetworks/peps.jl

```diff
@@ -0,0 +1,142 @@
+Optionally specify the link dimension with the keyword argument `linkdims`, which by default is 1.
+"""
+function PEPS(::Type{T}, sites::Matrix{<:Index}; linkdims::Integer=1) where {T<:Number}
+  Ny, Nx = size(sites)
+  tensor_grid = Matrix{ITensor}(undef, Ny, Nx)
+  # we assume the PEPS at least has size (2,2). Can generalize if necessary
+  @assert(Nx >= 2 && Ny >= 2)
+
+  lh = Matrix{Index}(undef, Ny, Nx - 1)
+  for ii in 1:(Nx - 1)
+    for jj in 1:(Ny)
+      lh[jj, ii] = Index(linkdims, "Lh,$jj,$ii")
+    end
+  end
+  lv = Matrix{Index}(undef, Ny - 1, Nx)
+  for ii in 1:(Nx)
+    for jj in 1:(Ny - 1)
+      lv[jj, ii] = Index(linkdims, "Lv,$jj,$ii")
+    end
+  end
+
+  # boundary cases
+  tensor_grid[1, 1] = emptyITensor(T, lh[1, 1], lv[1, 1], sites[1, 1])
+  tensor_grid[1, Nx] = emptyITensor(T, lh[1, Nx - 1], lv[1, Nx], sites[1, Nx])
+  tensor_grid[Ny, 1] = emptyITensor(T, lh[Ny, 1], lv[Ny - 1, 1], sites[Ny, 1])
+  tensor_grid[Ny, Nx] = emptyITensor(T, lh[Ny, Nx - 1], lv[Ny - 1, Nx], sites[Ny, Nx])
+  for ii in 2:(Nx - 1)
+    tensor_grid[1, ii] = emptyITensor(T, lh[1, ii], lh[1, ii - 1], lv[1, ii], sites[1, ii])
+    tensor_grid[Ny, ii] = emptyITensor(
+      T, lh[Ny, ii], lh[Ny, ii - 1], lv[Ny - 1, ii], sites[Ny, ii]
+    )
+  end
+
+  # inner sites
+  for jj in 2:(Ny - 1)
+    tensor_grid[jj, 1] = emptyITensor(T, lh[jj, 1], lv[jj, 1], lv[jj - 1, 1], sites[jj, 1])
+    tensor_grid[jj, Nx] = emptyITensor(
+      T, lh[jj, Nx - 1], lv[jj, Nx], lv[jj - 1, Nx], sites[jj, Nx]
+    )
+    for ii in 2:(Nx - 1)
+      tensor_grid[jj, ii] = emptyITensor(
+        T, lh[jj, ii], lh[jj, ii - 1], lv[jj, ii], lv[jj - 1, ii], sites[jj, ii]
+      )
+    end
+  end
+
+  return PEPS(tensor_grid)
+end
+
+PEPS(sites::Matrix{<:Index}, args...; kwargs...) = PEPS(Float64, sites, args...; kwargs...)
+
+PEPS(data::Array, size::Tuple{Int64,Int64}) = PEPS(reshape(data, size[1], size[2]))
+
+function randomizePEPS!(P::PEPS)
+  dimy, dimx = size(P.data)
+  for ii in 1:dimx
+    for jj in 1:dimy
+      randn!(P.data[jj, ii])
+      normalize!(P.data[jj, ii])
+    end
+  end
+end
+
+function Base.:+(A::PEPS, B::PEPS)
+  @assert(size(A.data) == size(B.data))
+  out = [t1 + t2 for (t1, t2) in zip(vcat(A.data...), vcat(B.data...))]
+  return PEPS(out, size(A.data))
+end
+
+function Base.:-(A::PEPS, B::PEPS)
+  @assert(size(A.data) == size(B.data))
+  out = [t1 - t2 for (t1, t2) in zip(vcat(A.data...), vcat(B.data...))]
+  return PEPS(out, size(A.data))
+end
+
+function Base.:*(c::Number, A::PEPS)
+  out = [c * t for t in vcat(A.data...)]
+  return PEPS(out, size(A.data))
+end
+
+function Base.:*(A::PEPS, B::PEPS)
+  @assert(size(A.data) == size(B.data))
+  out_scalar = 0
+  for (t1, t2) in zip(vcat(A.data...), vcat(B.data...))
+    out_scalar += ITensors.scalar(t1 * t2)
+  end
+  return out_scalar
+end
+
+get_prime_bonds(indices; ham=true) = ham ? indices : indices[1:(end - 1)]
```

Comment:
This definition assumes the last site is the site index which isn't always the case.

The generic way to handle this is to determine the link indices in a "network" style, where you determine the link indices based on the indices that are shared between neighbors in the network. This is what I implemented here:

https://github.com/ITensor/ITensorNetworkAD.jl/blob/c178a13c63f95f8bc51e322f16eb6ab217b7217b/src/ITensorNetworks/tensor_networks.jl#L641-L657

The way I would design it would be to have a simple `prime(P::PEPS)` function like this:
```julia
function ITensors.prime(P::PEPS)
  return PEPS(map(prime, P.data))
end
```
and then a separate function:
```julia
function ITensors.prime(::typeof(linkinds), P::PEPS)
  return PEPS(prime(linkinds, P.data))
end
```
Here it would call the generic `prime` function I linked to above. Unfortunately that doesn't specialize to the PEPS lattice structure (so searches through the entire network for neighbors instead of just the neighbors on the square lattice). 

I'm open to a different interface than `prime(linkinds, P)` for priming the bonds, this is just the interface I came up with in ITensors.jl for the MPS type.

Reply:
If I understand correctly, implementing this needs to change the PEPS struct. If that's the case, I think we'd better do that in a later PR.

Reply:
This wouldn't require changing the `PEPS` class, just changing the definition of `prime(::PEPS)` below to call the general `prime(tn)`/`prime(linkinds, tn)` function internally. I think this should be fixed because the `get_prime_bonds` function can be wrong if the index ordering isn't the same as the one assumed in `get_prime_bonds`.

Reply:
I see, done